firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/FirebaseAppDistribution.java

```diff
@@ -119,7 +134,107 @@ public void onFailure(@NonNull @NotNull Exception e) {
+    checkForUpdateHandler = HandlerCompat.createAsync(Looper.getMainLooper());
+
+    firebaseInstallationsApi
+        .getId()
+        .addOnSuccessListener(
+            new OnSuccessListener<String>() {
+              @Override
+              public void onSuccess(String fid) {
+                firebaseInstallationsApi
+                    .getToken(true)
+                    .addOnSuccessListener(
+                        new OnSuccessListener<InstallationTokenResult>() {
+                          @Override
+                          public void onSuccess(InstallationTokenResult installationTokenResult) {
+
+                            String authToken = installationTokenResult.getToken();
+                            String appId = firebaseApp.getOptions().getApplicationId();
+                            String apiKey = firebaseApp.getOptions().getApiKey();
+
+                            //make into method parameters
+                            FirebaseAppDistributionTesterApiClient client =
+                                new FirebaseAppDistributionTesterApiClient(fid, appId, apiKey);
+
+                            handleLatestReleaseFromClient(client, authToken);
+                          }
+                        });
+              }
+            })
+        .addOnFailureListener(
+            new OnFailureListener() {
+              @Override
+              public void onFailure(@NonNull Exception e) {
+                setCheckForUpdateTaskCompletionError(
+                    new FirebaseAppDistributionException(
+                        FirebaseAppDistributionException.Status.UNKNOWN));
+              }
+            });
+
+    return checkForUpdateTaskCompletionSource.getTask();
+  }
+
+  private void handleLatestReleaseFromClient(
+      FirebaseAppDistributionTesterApiClient client, String authToken) {
+    checkForUpdateExecutor.execute(
+        new Runnable() {
+          @Override
+          public void run() {
+            try {
+              AppDistributionRelease latestRelease = client.fetchLatestRelease(authToken);
+
+              Context context = firebaseApp.getApplicationContext();
+              long currentInstalledVersionCode = getInstalledAppVersionCode(context);
+
+              //move logic outside,
+              checkForUpdateHandler.post(
+                  new Runnable() {
+                    @Override
+                    public void run() {
+                      if (currentInstalledVersionCode
+                          < Long.parseLong(latestRelease.getBuildVersion())) {
+                        // latest release is newer
+                        checkForUpdateTaskCompletionSource.setResult(latestRelease);
+                      } else {
+                        checkForUpdateTaskCompletionSource.setResult(null);
+                      }
+                    }
+                  });
+            } catch (Exception e) {
```

Comment:
Can we use a more explicit Exception type here? https://source.android.com/setup/contribute/code-style#dont-catch-generic-exception